### ISSUE TITLE
Convert icon font to SVG: make icons screen-reader friendly and easier to update

### ISSUE THREAD

**Issue by JoshBarr** [wagtail/wagtail#1224]
We just had a user try to access the CMS from a tethered mobile device, and they were worried about all the "q" characters showing up in the admin interface before the fonts loaded... I think possibly their connection closed before the whole font downloaded, so they had a mix of icons and placeholder characters showing up in their interface.  

Is there a way we could more gracefully handle the flash of unstyled content, or preload the typefaces ahead of showing the UI?

**Comment by davecranwell** [comment]
The more I read up on icon fonts the more I conclude they're actually not that great, for precisely these brittleness issues. I'd much rather convert all icon use to SVG sprites.

**Comment by tmsndrs** [comment]
I agree. Inline SVG is much better. I've had a couple of sessions where the icon font just hasn't loaded at all (I didn't look into why at the time). 

Good SVG vs icon font write up here https://css-tricks.com/icon-fonts-vs-svg/

**Comment by davecranwell** [comment]
That's the exact article I had in mind.

**Comment by JoshBarr** [comment]
+1. We use SVG for pretty much everything.

**Comment by tomdyson** [comment]
@JoshBarr is this something that you or your team could help with?

**Comment by davecranwell** [comment]
The use of pseudo-elements is also a pain in the neck at the moment, due to the slightly unpredictable nature of EMs etc.

I'd imagine the fix for this might be to replace all uses of icon-specific classes with a template tag that outputs `<svg><use ...></svg>`, referencing a giant inline svg sprite at the top of the page.

Does this align with your approach @JoshBarr ?

**Comment by gasman** [comment]
If we're going to be rethinking our approach to icons, then let me put in a few wishlist requests...

* If adding a new icon is going to involve updating a central resource (e.g. font / sprite), document the process for doing that
* Make it possible for end-users to define their own icons in addition to wagtailadmin's stock set, for use in StreamField blocks or custom (hook-defined) menu items

**Comment by davecranwell** [comment]
Adding icons to the global set could be done with a hook which adds a user's own svg sprite inline to the top of the page. Our proposed icon template tag would continue to work as normal, because the user could just reference the IDs within their custom svg instead of the core one.

**Comment by JoshBarr** [comment]
We've been using a CSS file with base64 encoded SVGs through Grunticon, like this: [Kiwibank sprite sheet](http://kiwibank.co.nz/assets-personal-banking/css/icons.data.svg.css)

If you're not worried about IE8 support (falls back to PNGs), then inline is possibly better and requires less ceremony than Grunticon to get up and running. 

@tomdyson sure thing, I think we could swap out the icons to SVG for you.

**Comment by davecranwell** [comment]
IE8 is definitely out of the picture. The sprite sheet solution avoids the need to use rather ugly <svg> inline, but does mean we're less able to style or animate SVG within the page, but we've not found a need to do that yet anyway. A sprite sheet is perhaps a more linear progression from what we have currently too.

So long as we're replacing pseudo-elements with actual elements, `span`s or whatever, it's an improvement.

**Comment by JoshBarr** [comment]
Hey @davecranwell, I've just exported the wagtail `.ttf` font into SVGs to start the move away from icon fonts. Found a great thread on automating glyph extraction: http://stackoverflow.com/questions/18113478/extracting-svg-from-font-awesome 

Before I run off and replace a couple of hundred icon references, I thought I'd check in with you about your preferred approach. Would you like to use inline `<svg>` tags with `#fragment-identifiers`? 

![screen shot 2015-05-04 at 11 25 58 pm](https://cloud.githubusercontent.com/assets/1715456/7452699/1921d0e8-f2b5-11e4-8a4a-17c8ee099488.png)

**Comment by davecranwell** [comment]
Sorry, I should have mentioned this before, but the wagtail font is created with an Icomoon project. In the same folder as the fonts is a project json file which you can import into icomoon with "import project" here https://icomoon.io/app/#/projects, then choose the first button at the bottom "Generate SVG/PDF/PNG" https://icomoon.io/app/#/select/image 

Inline SVG with fragment identifiers seems to one of many recommended approaches, but seems sensible, unless you have another preference? If we did use fragments we'd probably want to wrap the placement with a django template tag to make it slightly less verbose in the template files themselves.

**Comment by JoshBarr** [comment]
Yeah, it feels like a pretty good approach so far, as the SVGs pick up the currentColor of the text, meaning they behave a lot like the icon font.  

I've set up a template tag with this nice simple syntax: `{% icon 'arrow-right' %}`, but you can also overload it with extra classes if you need to `{% icon 'arrow-right' 'large' %}`

I have a feeling this may take several hours!

**Comment by allcaps** [comment]
I will give this a go. My goals are:

- Current Wagtail icons => svg symbols.
- use tag
- template tag
- Way to hook in extra icons (your app or third party).
- Way to override some icons (customisation)
- Way to override all icons (icon packs)
- Documentation
- See if SVG icons need to be in the base template or can live in a external file/view.

**Comment by thibaudcolas** [comment]
Closing this in favour of #6107. This issue now feels quite out of date, as most of the problems stated there are fixed by #5359, and further #4821.
